src/string/memccpy.c

```diff
@@ -7,21 +7,32 @@
 #define HIGHS (ONES * (UCHAR_MAX/2+1))
 #define HASZERO(x) ((x)-ONES & ~(x) & HIGHS)
 
-void *memccpy (void *__restrict dest : itype(__restrict _Array_ptr<void>) byte_count(n),
-               const void *__restrict src : itype(__restrict _Array_ptr<const void>) byte_count(n),
+void *memccpy (void *__restrict dest : itype(__restrict _Array_ptr<void>)
+                                       bounds((_Array_ptr<unsigned char>)dest,
+                                              (_Array_ptr<unsigned char>)dest + arg_n),
+               const void *__restrict src : itype(__restrict _Array_ptr<const void>)
+                                            bounds((_Array_ptr<const unsigned char>)src,
+                                                   (_Array_ptr<const unsigned char>)src + arg_n),
                int c,
-               size_t n)
-  : itype(_Array_ptr<void>) byte_count(n)
+               size_t arg_n)
+  : itype(_Array_ptr<void>) byte_count(arg_n)
 _Checked
 {
-        _Array_ptr<unsigned char> d : count(n) = (_Array_ptr<unsigned char>)dest;
-        _Array_ptr<const unsigned char> s : count(n) = (_Array_ptr<const unsigned char>)src;
+	size_t n = arg_n;
+        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)dest,
+                                             (_Array_ptr<unsigned char>)dest + arg_n)
+                                    = (_Array_ptr<unsigned char>)dest;
+        _Array_ptr<const unsigned char> s : bounds((_Array_ptr<const unsigned char>)src,
+                                                   (_Array_ptr<const unsigned char>)src + arg_n)
+                                          = (_Array_ptr<const unsigned char>)src;
 
 	c = (unsigned char)c;
 #ifdef __GNUC__
 	typedef size_t __attribute__((__may_alias__)) word;
-	_Array_ptr<word> wd : count(n) = 0;
-	_Array_ptr<const word> ws : count(n) = 0;
+	_Array_ptr<word> wd : bounds((_Array_ptr<unsigned char>)dest,
+                                     (_Array_ptr<unsigned char>)dest + arg_n) rel_align(unsigned char) = 0;
+	_Array_ptr<const word> ws : bounds((_Array_ptr<const unsigned char>)src,
+                                           (_Array_ptr<const unsigned char>)src + arg_n) rel_align(unsigned char) = 0;
```

Comment:
Could you please add comments on why `rel_align` is needed here?

Reply:
As `src` and `dest` are pointers to unsigned char, it is possible that they may not be aligned to a word boundary (here, `word` is typedef'd to `size_t`). When `ws` or `wd` are dereferenced, they will read `sizeof(word)` bytes because they are of type `word`. We need to add `rel_align(unsigned char)` to inform the compiler that the bounds for `ws` and `wd` are aligned to `unsigned char`, whereas `ws` and `wd` dereference a `word` at a time. When the compiler is thus informed, it emits a runtime bounds check as follows: `ws + sizeof(word) - 1 < upper-bound`. Please see section 3.9.2 of the spec for further details. Note that if `rel_align` were not specified, the compiler will emit a runtime bounds check as: `ws < upper-bound`. This would clearly not prevent an oob access.
PS: The implementation of the runtime check emission is yet to happen - the description above is expected behavior.